activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, options)
+      @model = model
+      @inserts = inserts
+      @options = options
+    end
+
+    def execute
+      return ActiveRecord::Result.new([], []) if inserts.empty?
+
+      connection.exec_query to_sql, "Bulk Insert"
+    end
+
+    def to_sql
+      raise ArgumentError, "inserts is empty" if inserts.empty?
+
+      connection.build_insert_all_sql(
```

Comment:
Right, should flip it around, such that the SQL building happens here, passing in the connection if needed.